src/string/src/lib.rs

```diff
@@ -12,13 +14,41 @@ use core::cmp;
 use core::usize;
 
 #[no_mangle]
-pub extern "C" fn memccpy(s1: *mut c_void, s2: *const c_void, c: c_int, n: usize) -> *mut c_void {
-    unimplemented!();
+pub unsafe extern "C" fn memccpy(
+    dest: *mut c_void,
+    src: *const c_void,
+    c: c_int,
+    n: usize,
+) -> *mut c_void {
+    use compiler_builtins::mem::memcpy;
+    use core::mem;
+    let dest = dest as *mut u8;
+    let to = memchr(src, c, n);
+    if to as usize == 0 {
+        return to;
+    }
+    let src = src as *mut u8;
+    let dist = ((to as usize) - (src as usize)) / mem::size_of::<u8>();
+    if memcpy(dest, src, dist) as usize > 0 {
```

Comment:
https://github.com/rust-lang-nursery/compiler-builtins/blob/master/src/mem.rs#L8-L19 it looks like the compiler builtin version of `memcpy` cannot return NULL. So this will always be true.

Reply:
I believe it just returns dest, so if `dest` is `null`, it will return `null`. I'm more than happy to change it though

Reply:
Ah, makes sense. Last readability nit: `!memcpy(dest, src, dist).is_null()`

Reply:
ok, from testing, it won't cause a segfault, but the output will be incorrect

Reply:
I could change it to `if memcpy(dest, src, dist).is_null() { ... }` in the spirit of one of the above comments though.

Reply:
:+1: looks great! Thanks for the work on this!

Reply:
No worries!